### ISSUE TITLE
Custom config file not being loaded

### ISSUE THREAD

**Issue by SageEx** [Immington-Industries/way-cooler#245]
Hi everyone,
First of all, thanks for this project.
I'm currently having trouble with way-cooler not loading my config file.
I installed way-cooler from AUR : ``way-cooler-git``.

config file loation: ``~/.config/way-cooler/init.lua``
env $XDG_CONFIG_HOME: ``/home/sageex/.config``

I have changed the default terminal, the mod key etc. and the changes are not loading.
Is there a cli flag to load the init file ?

Also, is there some documentation for the init file ? Like man pages ?

**Comment by Timidger** [comment]
Hey there. 

If you could provide a log that would be very helpful (you can do that by redirecting stderr to a file, eg `way_cooler 2> log.txt`).

Another thing -  we recently merged in a change that has the default configuration file used if the user defined one has an error. It's possible your custom configuration file has an error in it. Can you post that as well?

**Comment by SageEx** [comment]
Hi, so here are the asked for files.
[log.txt](https://github.com/Immington-Industries/way-cooler/files/745411/log.txt)

Here's the init.lua file
```lua
way_cooler.background = 0x5E4055

mod = "Super"
local key = config.key -- Alias key so it's faster to type

way_cooler.terminal = "st" -- Use the terminal of your choice
way_cooler.background = "0x111111"
way_cooler.bar = "polybar"

way_cooler.gap_size = 3 -- The width of gaps between windows in pixels
way_cooler.border_size = 1 -- The width of the borders between windows
way_cooler.border_color = 0x386890 -- The color of the borders
way_cooler.active_border_color = 0x57beb9 -- Color of active container borders

local keys = {
  -- Open dmenu
  key({ mod }, "r", "launch_dmenu"),
  -- Open terminal
  key({ mod }, "return", "st"),

  -- Lua methods can be bound as well
  key({ mod, "Shift" }, "h", function () print("Hello world!") end),

  -- Some Lua dmenu stuff
  key({ mod }, "l", "dmenu_eval"),
  key({ mod, "Shift" }, "l", "dmenu_lua_dofile"),

  -- Move focus
  key({ mod }, "left", "focus_left"),
  key({ mod }, "right", "focus_right"),
  key({ mod }, "up", "focus_up"),
  key({ mod }, "down", "focus_down"),

  -- Move active container
  key({ mod, "Shift" }, "left", "move_active_left"),
  key({ mod, "Shift" }, "right", "move_active_right"),
  key({ mod, "Shift" }, "up", "move_active_up"),
  key({ mod, "Shift" }, "down", "move_active_down"),

  -- Split containers
  key({ mod }, "h", "split_horizontal"),
  key({ mod }, "v", "split_vertical"),
  key({ mod }, "e", "horizontal_vertical_switch"),
  key({ mod }, "f", "fullscreen_toggle"),
  key({ mod, "Shift" }, "q", "close_window"),
  key({ mod, "Shift" }, "space", "toggle_float_active"),
  key({ mod }, "space", "toggle_float_focus"),
  key({ mod, "Shift" }, "r", "way_cooler_restart")

  -- Quitting way-cooler is hardcoded to Alt+Shift+Esc.
  -- This my be modifiable in the future
}

-- Add Mod + X bindings to switch to workspace X, Mod+Shift+X send active to X
for i = 1, 9 do
  table.insert(keys,
               key({ mod }, tostring(i), "switch_workspace_" .. i))
  table.insert(keys,
               key({ mod, "Shift" }, tostring(i), "move_to_workspace_" .. i))
end

-- Register the keybindings.
for _, key in pairs(keys) do
    config.register_key(key)
end

-- Register the mod key to also be the mod key for mouse commands
config.register_mouse_modifier(mod)

function cleanup_background()
  os.execute("pkill way-cooler-bg")
end


-- Execute some code after Way Cooler is finished initializing
function way_cooler_init()
  local status = os.execute("which way-cooler-bg 2>/dev/null")
  if not status then
    print "Could not find way-cooler-bg! Please install it"
  else
    os.execute("way-cooler-bg " ..  way_cooler.background .. " &")
  end
end

--- Execute some code when Way Cooler restarts
function way_cooler_restart()
  cleanup_background()
end

function way_cooler_terminate()
  cleanup_background()
end


way_cooler.on_restart(way_cooler_restart)
way_cooler.on_terminate(way_cooler_terminate)


util.exec.spawn_once("polybar topbar")
util.exec.spawn_once("way-cooler-bar")
util.exec.spawn_once("st")

-- To add your own Lua files:
-- require("my-config.lua") -- Or use utils.hostname

-- !! Do not place any code after this comment.
-- !! way-cooler and plugins may insert auto-generated code.
```

Note that I have changed the deafult Mod to Super key, but the log file shows them all bind to the alt key. Also, if a user keybinding conflicts with a system keybinding, the new  keybinding is blocked ( in the stderr).

**Comment by SageEx** [comment]
Also, the following settings seem to do nothing,
```lua
way_cooler.terminal = "st" -- Use the terminal of your choice
way_cooler.background = "0x111111"
way_cooler.bar = "polybar"

way_cooler.gap_size = 3 -- The width of gaps between windows in pixels
way_cooler.border_size = 1 -- The width of the borders between window
```
In the attached log file, it said ``init.lua`` read successfully, but these settings are still the same.

**Comment by Timidger** [comment]
Ah, ok I have found the error.

Looks like there is a bug in the error reporting for the custom lua config file. Eventually we'll have a little notification pop up when there is an issue so it is more explicit where something went wrong to the user. I'll need to tweak the current implementation though so that we log this failure properly.

The bug is on the line where you define `key({ mod }, "return", "st"),`. The last argument to the key function needs to be either a defined Lua function (e.g something the user has defined) or the name of a special command for Way Cooler. `launch_terminal` is the command that is simply a wrapper around the call to whatever you defined `way_cooler.terminal` to be. So the line should read `key({ mod }, "return", "launch_terminal")`

I admit this is very confusing, and this would have been much easier to understand if we had sufficient documentation. We don't have man pages or anything like that just yet, because the state of the configuration options is in a bit of a flux. I'm working on a patch right now that will seriously alter the configuration file (and be backward incompatible). Once that is done though, and I'm confident the new format is what I want, I can begin writing proper documentation on both the configuration file and the what the user can actually do with the Lua integration.

**Comment by Timidger** [comment]
Oh and another issue I found -- you're using `config.exec.spawn_once` but we don't yet include those libraries in the configuration file :(. The comment talking about them is out of date, you'll need to wrap those in the `way_cooler_init` function like `way-cooler-bg` is. This also **will** be fixed soon in a configuration overhaul patch.

**Comment by SageEx** [comment]
Thanks for this, I'll be checking this out.

I'd be very interested in working on proper user documentation if you need help.

**Comment by SageEx** [comment]
Hi again,
The current issue seems to be fixed, my config file is being read.

How do I bind custom keybindings to programs ?
I came up with this, but it errors out. The function is being executed when I reload waty-cooler, and then way-cooler falls back to the system init.lua.

```lua
function launch_program(prog_name)
    os.execute(prog_name .. " &")
end
```
and in keys section
```lua
   key({ mod }, "d", launch_program("dmenu_run")),
```

Any idea how to fix this ?

**Comment by SageEx** [comment]
Solved it for the time being,
```lua
  key({ mod }, "d", function () launch_program("dmenu_run") end),
```

**Comment by SageEx** [comment]
Shall I close this issue for now ?
Or would it be better to close it after the overhaul ?

**Comment by Timidger** [comment]
>I'd be very interested in working on proper user documentation if you need help.

That would be excellent. If you are interested in helping out though, I suggest waiting until I change the configuration file in a pr. 

Feel free to also comment on the pr of the new format has any problems you forsee. I'm hoping to keep the configuration file relatively stable even pre 1.0.

> Solved it for the time being, <snippet>

Yeah that's how you'd have to do it, because otherwise the function will be evaluated and the output will be used as a value. 

> Shall I close this issue for now ?

Yeah, I'll close it. I have opened a relevant issue to fix this log problem. Thanks for bringing this to my attention :-)
